bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -538,52 +589,73 @@ protected String executeSingle(ResourceBundle language, String[] labelFragments,
      * @param language Language information that is used for matching
      * @param labelFragments label fragments that are used to match an item's label.
      *            For a positive match, the item's label has to contain every fragment - independently of their order.
-     *            They are treated case insensitive.
+     *            They are treated case-insensitive.
      * @param commandType optional command type that all items have to support.
      *            Provide {null} if there is no need for a certain command to be supported.
      * @return All matching items from the item registry.
      */
     protected List<Item> getMatchingItems(ResourceBundle language, String[] labelFragments,
             @Nullable Class<?> commandType) {
-        List<Item> items = new ArrayList<>();
-        Map<Item, List<Set<String>>> map = getItemTokens(language.getLocale());
-        for (Item item : map.keySet()) {
-            for (Set<String> parts : map.get(item)) {
-                boolean allMatch = true;
-                for (String fragment : labelFragments) {
-                    if (!parts.contains(fragment.toLowerCase(language.getLocale()))) {
-                        allMatch = false;
+        Set<Item> items = new HashSet<>();
+        Set<Item> exactMatchItems = new HashSet<>();
+        Map<Item, List<List<List<String>>>> map = getItemTokens(language.getLocale());
+        for (Entry<Item, List<List<List<String>>>> entry : map.entrySet()) {
+            Item item = entry.getKey();
+            for (List<List<String>> parts : entry.getValue()) {
+                boolean exactMatch = false;
+                logger.trace("Checking tokens {} against the item tokens {}", labelFragments, parts);
+                List<String> lowercaseLabelFragments = Arrays.stream(labelFragments)
+                        .map(lf -> lf.toLowerCase(language.getLocale())).toList();
+                List<String> unmatchedFragments = new ArrayList<>(lowercaseLabelFragments);
+                for (List<String> segment : parts) {
+                    if (segment.equals(lowercaseLabelFragments)) {
+                        exactMatch = true;
+                        unmatchedFragments.clear();
                         break;
                     }
+                    if (!unmatchedFragments.isEmpty()) {
+                        segment.stream().filter(unmatchedFragments::contains).forEach(unmatchedFragments::remove);
+                    }
```

Comment:
```suggestion
                    unmatchedFragments.removeAll(segment);
```

Wouldn't this also work?